### ISSUE TITLE
Incorrect packet type?

### ISSUE THREAD

**Issue by Affector** [micheloosterhof/kippo#11]
I have switched from the Desaster-Kippo to your fork and now i'm getting errors when clients connect to the honeypot. I'm a not a true crypto-artist, but this looks to me a incorrect handshake..?

2015-02-18 13:25:40+0100 [HoneyPotTransport,0,127.0.0.2] Remote SSH version: SSH-2.0-OpenSSH_6.0p1 Debian-4+deb7u2
2015-02-18 13:25:40+0100 [HoneyPotTransport,0,127.0.0.2] kex alg, key alg: diffie-hellman-group-exchange-sha1 ssh-rsa
2015-02-18 13:25:40+0100 [HoneyPotTransport,0,127.0.0.2] outgoing: aes128-ctr hmac-md5 none
2015-02-18 13:25:40+0100 [HoneyPotTransport,0,127.0.0.2] incoming: aes128-ctr hmac-md5 none
2015-02-18 13:25:40+0100 [HoneyPotTransport,0,127.0.0.2] Got remote error, code 2
        reason: Protocol error: expected packet type 31, got 20
2015-02-18 13:25:40+0100 [HoneyPotTransport,0,127.0.0.2] connection lost

I got this error with two different clients. It this a client-side error?

**Comment by micheloosterhof** [comment]
I checked myself and I use the exact same client to connect (OpenSSH_6.0p1 Debian-4+deb7u2, OpenSSL 1.0.1e 11 Feb 2013) and I can connect without issue. This means it perhaps has something to do with your configuration. Which two clients did you use, both OpenSSH on Debian? Did it work with desaster's on the same machine?

Do you use custom encryption settings for ssh client? (in your ~/.ssh/config or /etc/ssh/ssh_config)?

I'm not seeing any KEXINIT messages in your log fragment. Are you using the latest 'master' branch?

It could maybe be the diffie-helpman-group-exchange-sha1, desaster's kippo doesn't have it. Do you have the /etc/ssh/moduli file and is it readable by the kippo user id?

**Comment by Affector** [comment]
It works with Desaster's version on the same machine. I have made a copy of /etc/ssh/moduli and added that to my Kippo-install, so i don't have to give Kippo read-rights on the original /etc/ssh/moduli.

Here is a complete request i found in the logfile..
2015-02-18 13:36:57+0100 [kippo.core.ssh.HoneyPotSSHFactory] New connection: 127.0.0.4:54461 (149.210.177.2:2222) [session: 1]
2015-02-18 13:36:57+0100 [HoneyPotTransport,1,127.0.0.4] KEXINIT: client supported key exchange: ['ecdh-sha2-ni..<shortened for readability>
2015-02-18 13:36:57+0100 [HoneyPotTransport,1,127.0.0.4] KEXINIT: client supported public keys: ['ecdsa-sha2-ni..<shortened for readability>
2015-02-18 13:36:57+0100 [HoneyPotTransport,1,127.0.0.4] KEXINIT: client supported encryption: ['aes128-ctr', ..<shortened for readability>
2015-02-18 13:36:57+0100 [HoneyPotTransport,1,127.0.0.4] KEXINIT: client supported MAC: ['hmac-md5..<shortened for readability>
2015-02-18 13:36:57+0100 [HoneyPotTransport,1,127.0.0.4] KEXINIT: client supported compression: ['none', 'zlib@openssh.com', 'zlib']
2015-02-18 13:36:57+0100 [HoneyPotTransport,1,127.0.0.4] KEXINIT: client supported lang: ['']
2015-02-18 13:36:57+0100 [HoneyPotTransport,1,127.0.0.4] Remote SSH version: SSH-2.0-OpenSSH_6.1
2015-02-18 13:36:57+0100 [HoneyPotTransport,1,127.0.0.4] kex alg, key alg: diffie-hellman-group-exchange-sha1 ssh-rsa
2015-02-18 13:36:57+0100 [HoneyPotTransport,1,127.0.0.4] outgoing: aes128-ctr hmac-md5 none
2015-02-18 13:36:57+0100 [HoneyPotTransport,1,127.0.0.4] incoming: aes128-ctr hmac-md5 none
2015-02-18 13:36:57+0100 [HoneyPotTransport,1,127.0.0.4] Got remote error, code 2
        reason: Protocol error: expected packet type 31, got 20

**Comment by micheloosterhof** [comment]
I think it's related to the 'moduli' file. I currently only test for file
existance, not for read access.
If you copied the moduli file, did you also change the path in
kippo/core/ssh.py?

**Comment by Affector** [comment]
Yes, i have also changed the path in the kippo/core/ssh.py. Furthermore, i saw this strange string in /kippo/core/sshserver.py on line 60: 
```python
if remoteVersion not in self.supportedVersions
```
But self *(HoneyPotTransport when called)* has no attribute **'supportedVersions'**

**Comment by micheloosterhof** [comment]
I've just pushed a change with better exception handling for the moduli file. If you can pull that change, it may give some more insight.

**Comment by Affector** [comment]
I have pulled you change, but it's still giving the same error. Maybe the problem lies in the handling of the handshake, which is done by Twistd?

**Comment by micheloosterhof** [comment]
Okay, diving a little deeper, the type 20 is a rekey-request. 
http://www.snailbook.com/faq/no-rekeying.auto.html
That's expected to be done after a lot of data transfer/time, but it's triggered immediately here. 

Do you have any custom options in /etc/ssh/ssh_config or your ~/.ssh/ssh_config?

**Comment by micheloosterhof** [comment]
What is the Twisted version you are using? 
There are some updates in the underlying code around rekeying. I developed on 12.0

**Comment by Affector** [comment]
I got it working !! I have reinstalled Twisted *(and upgraded to version 15)* and that fixed my problem. My honeypot-setup resides behind a SSL-Accelerator which also does DPI, but i have especially made a tunnel to my setup. The server-administrator of the host-server *(the honeypot is a VM)* could have changed something for the SSL/Twisted.. but he doesn't remember. The SSL-offloading still works, so everybody's happy !!

This issue can be closed
